### DISCUSSION TITLE
How to add a new Qt designer form class in Ripes

### DISCUSSION THREAD

**Discussion by Y1annis** [mortbopet/Ripes#244] [General]
Hello, I have been working on Ripes for about a month now mainly on adding more functionalities to the Cache. Now I am trying to add a class to display some statistics that I calculated in cachesim file. I created another tab next to L1 Data cache and I will try to design it there. My problem is that when I create the new Class the .cpp file cant find the ui_stats.h file. How can I help the .cpp file to find the ui file in the project?

Thank you for your time.

**Comment by mortbopet** [comment]
Hi @Y1annis - awesome to hear, looking forward to seeing what you've done. There's various Qt guides for this, but what I've always done is to select new file -> Qt -> Qt designer form class. Then It creates the .h, .cpp, and .ui files all at once. Assuming you then place them alongside all the other source files, the `ui_*.h` file should get correctly generated.

**Reply by Y1annis** [reply]
Thank you for your answer @mortbopet. That is what I am doing, I looked at a lot of tutorials but for some reason it says that the cpp file cant find the ui_*.h file. Do I need to change anything in the Cmakelist files?

**Reply by mortbopet** [reply]
If you place the .ui file in the [src](https://github.com/mortbopet/Ripes/tree/master/src) folder then it should automatically get generated (due to [this](https://github.com/mortbopet/Ripes/blob/master/src/CMakeLists.txt#L18) glob pattern).

**Reply by Y1annis** [reply]
![image](https://user-images.githubusercontent.com/109152409/199770293-6f9d81ab-5c94-40f8-bde3-7f3ea652390f.png)

I get this warning when I create the new Qt designer form class, then I run Cmake so the files are visible.

And then I get this error in the cpp file 
![image](https://user-images.githubusercontent.com/109152409/199771367-b90bd964-5798-447a-af9f-f9bacd4f74dd.png)

They are all in the directory Ripes/src/Cachesim

**Reply by mortbopet** [reply]
1. A second level of finicky-ness in these UI files and Qt designer is that one must also write the `ripes::` namespace within the widget name itself in Qt designer - this isn't super obvious from Qt's side. E.g. in designer, if you widget is named `MyWidget`, change the name to `ripes::MyWidget`. Alternatively, you can edit the UI file manually and add the prefix, [as seen here](https://github.com/mortbopet/Ripes/blob/master/src/addressdialog.ui#L3-L4).
2. After running CMake and trying to build the project, is the `ui_cachestats.h` file _anywhere_ in your hierarchy?
3. Can you post your `.ui` file here?

**Reply by Y1annis** [reply]
I have added ripes:: in the name of my Widget, then I ran Cmake again but the problem is still there. Here is my .ui file:

`<?xml version="1.0" encoding="UTF-8"?>
<ui version="4.0">
 <class>Ripes::CacheStats</class>
 <widget class="QWidget" name="Ripes::CacheStats">
  <property name="geometry">
   <rect>
    <x>0</x>
    <y>0</y>
    <width>400</width>
    <height>300</height>
   </rect>
  </property>
  <property name="windowTitle">
   <string>Form</string>
  </property>
 </widget>
 <resources/>
 <connections/>
</ui>`

I didn't add anything in it yet because I am still trying to get it to work.

Thank you again for your time

**Reply by mortbopet** [reply]
And is `ui_cachestats.h` available anywhere in the build folder?

**Reply by Y1annis** [reply]
I just found the problem. In the build folder in cachesim there is cachesim_lib_autogen. When I created the new form class I was selecting cachesim_lib. I just created a new form class and it works now so I can finally start working on the new Class.

Thanks a lot for your help!

**Comment by Y1annis** [comment]
@mortbopet Hello again

I am working on my new CacheStatistics class and I am trying to model it after cacheplotwidget. I have created my functions and the new CacheWidget but when I try to run the project I get this error. Do you know what causes it?

I also added a new tab in CachetabWidget with my new ui.

![image](https://user-images.githubusercontent.com/109152409/202918241-c8fa9bb9-d0b9-49cb-b02d-e05fab8b3035.png)


Thank you for your time
